content/posts/running-flutter-integration-tests-in-docker/index.md

```diff
@@ -0,0 +1,116 @@
++++
+date = "2021-12-28T09:49:36+07:00"
+author = "phamnhuvu-dev"
+description = "Why and How we run integration tests in Docker"
+title = "Running Flutter integration tests in Docker"
+categories = ["DevSecOps", "Testing"]
+tags = ["docker", "integration-test", "flutter", "dart"]
+slug = "running-flutter-integration-tests-in-docker"
++++
+# Running Flutter integration tests in Docker
+
+## Prerequisite
+- Linux OS
+- Installing:
+    - Flutter
+    - Docker
+    - Docker compose
+    - KVM https://help.ubuntu.com/community/KVM/Installation
+
+## Technologies
+
+- Flutter:
+    - Google's SDK for crafting beautiful, fast user experiences for mobile, web, and desktop from a single codebase. Flutter works with existing code, is used by developers and organizations around the world, and is free and open source.
+    - https://github.com/flutter/flutter
+
+- Docker:
+    - An open platform for developing, shipping, and running applications
+    - https://docs.docker.com/get-started/overview/
+
+- Android Emulator container:
+    - Allowing you to find and run the right version of the Emulator without the headache of dependency management, which makes it easy to scale automated tests as part of a CI/CD system without the upkeep cost of a physical device farm
+    - https://github.com/google/android-emulator-container-scripts
+
+## The criteria that we use Docker for running automation tests
+
+- Scaling tests:
+    - Docker provides the ability to package and run an application in a loosely isolated environment, so we can run multiple e2e tests parallel.
+
+- Reusing:
+    - Our Dockerfile file and docker-compose.yml file is simple because of reusing container from the community.
+
+- Consistent:
+    - Allowing developers to work in standardized environments using local containers which provide applications and services
+
+- Responsive:
+    - Docker containers can run on a developer’s local laptop and on cloud providers.
+
+## Explanation
+
+- We create Dockerfile for the flutter-app service because we need to copy our source code to the container to build then run tests
+```
+FROM cirrusci/flutter:2.8.1
+
+RUN sdkmanager "build-tools;29.0.2"
+
+RUN flutter precache --android --no-web --no-ios --no-universal
+
+COPY pubspec.* .
+
+RUN --mount=type=cache,sharing=locked,target=/flutter flutter pub get
+
+WORKDIR /project
+
+COPY . .
+```
+
+- To run the Flutter integration test we need a process for Android Emulator and a process for Flutter drive. That is why we have 2 services is `flutter-app` and `android-emulator` in `docker-compose.yml` file.
```

Comment:
- To run the Flutter integration test we need a process for Android Emulator and a process for Flutter drive. That is why we have 2 services: `flutter-app` and `android-emulator` in `docker-compose.yml` file.